src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/14.diff.sql

```diff
@@ -0,0 +1,542 @@
+      CONSTRAINT UQ_ResourceChangeType_Name UNIQUE NONCLUSTERED 
+     (
+         Name ASC
+     )WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
+     ) ON [PRIMARY]
+
+     SET IDENTITY_INSERT [dbo].[ResourceChangeType] ON 
+
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+
+     SET IDENTITY_INSERT dbo.ResourceChangeType OFF
+END
+GO
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE OR ALTER PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId. The start id is inclusive.
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges
+    @startId bigint, 
+    @pageSize int
+AS
+BEGIN
+
+    SET NOCOUNT ON;
+
+    SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
+
+    SELECT TOP(@pageSize) c.Id
+      ,c.Timestamp
+      ,c.ResourceId
+      ,c.ResourceTypeId
+      ,t.Name AS ResourceTypeName
+      ,c.ResourceVersion
+      ,c.ResourceChangeTypeId
+      FROM dbo.ResourceChangeData c
+      LEFT JOIN dbo.ResourceType t ON c.ResourceTypeId = t.ResourceTypeId
+    WHERE c.Id >= @startId ORDER BY c.Id ASC
```

Comment:
Is it possible for some events to fail to be delivered but others to succeed?  What happens then?  Do we block all processing when we hit a failure ?

Reply:
We submit the change feed to the Event Grid API endpoint and ensure that a set of changes are reached in the Event Grid. A particular event is delivered to the subscriber or not is out of the scope. But, Event Grid ensures reliable delivery of events to the destination Endpoints.

Reply:
Can the call to Event Grid succeed for some events but not others?

Reply:
No, that can not happen.